File: ai-service/app/sensitive_data_detector.py
"""
Sensitive Data Detector (NLP Module)
Detects PII and confidential information using NLP entity patterns.
Complements the regex-based detection in the Java backend.
"""

import re
from typing import List, Tuple


class SensitiveDataDetectorNLP:
    """
    NLP-enhanced sensitive data detection.
    Focuses on contextual patterns that regex alone might miss.
    """

    # Contextual patterns (regex + surrounding context)
    CONTEXTUAL_PATTERNS = [
        # SSN in context
        (r'(social\s+security|ssn|ss#?)\s*[:\-]?\s*\d{3}[\-\s]?\d{2}[\-\s]?\d{4}', "ssn_in_context"),
        # Bank account in context
        (r'(account\s*(number|#|no\.?))\s*[:\-]?\s*\d{8,17}', "bank_account_in_context"),
        # Routing number in context
        (r'(routing\s*(number|#|no\.?)|aba)\s*[:\-]?\s*\d{9}', "routing_number_in_context"),
        # Credit card with context
        (r'(card\s*(number|#)|credit\s+card|visa|mastercard|amex)\s*[:\-]?\s*\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4}',
         "credit_card_in_context"),
        # API keys / tokens
        (r'(api[_\s\-]?key|access[_\s\-]?token|secret[_\s\-]?key|bearer)\s*[:\=]\s*\S{16,}', "api_key_in_context"),
        # Passwords
        (r'(password|passwd|pwd|pass)\s*[:\=]\s*\S{4,}', "password_in_context"),
        # Internal documents
        (r'\b(confidential|internal\s+only|restricted|classified|proprietary)\b', "confidential_marker"),
    ]

    def detect(self, text: str) -> Tuple[float, List[str]]:
        """
        Detect sensitive data with contextual NLP patterns.
        Returns: (score 0-100, list of signal names)
        """
        if not text:
            return 0.0, []

        signals = []
        score = 0.0

        for pattern, signal in self.CONTEXTUAL_PATTERNS:
            if re.search(pattern, text, re.IGNORECASE):
                if signal not in signals:
                    signals.append(signal)
                    score += 20

        # Multiple sensitive data types = higher risk
        if len(signals) >= 3:
            score += 15

        return min(score, 100), signals

File: backend/src/main/java/com/emailguardian/service/SensitiveDataDetector.java
package com.emailguardian.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.regex.*;

/**
 * Fast-layer sensitive data detection using regex patterns.
 * Detects: SSN, bank accounts, credit cards, API keys, passwords, financial data.
 * Runs entirely locally — no AI service dependency.
 */
@Service
@Slf4j
public class SensitiveDataDetector {

    // Regex patterns for sensitive data types
    private static final Map<String, Pattern> PATTERNS = Map.of(
        "ssn", Pattern.compile("\\b\\d{3}-\\d{2}-\\d{4}\\b"),
        "credit_card", Pattern.compile("\\b(?:\\d{4}[- ]?){3}\\d{4}\\b"),
        "bank_account", Pattern.compile("\\b\\d{8,17}\\b"),
        "routing_number", Pattern.compile("\\b0[0-9]{8}\\b|\\b[0-3]\\d{8}\\b"),
        "api_key", Pattern.compile("(?i)(api[_-]?key|apikey|access[_-]?token|secret[_-]?key|sk-)[\\s:=]+['\"]?[A-Za-z0-9_\\-]{16,}"),
        "password", Pattern.compile("(?i)(password|passwd|pwd)[\\s:=]+['\"]?[^\\s'\"]{4,}"),
        "email_address_bulk", Pattern.compile("(?:[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}[,;\\s]+){3,}")
    );

    // Keywords indicating financial/confidential data
    private static final List<String> FINANCIAL_KEYWORDS = List.of(
        "bank account", "routing number", "wire transfer", "account number",
        "credit card", "social security", "tax id", "ein",
        "confidential", "internal only", "do not share", "classified"
    );

    /**
     * Analyze email body for sensitive data patterns.
     * @return score 0-100 and list of detected signals
     */
    public SensitiveDataResult analyze(String emailBody, String subject) {
        if (emailBody == null || emailBody.isEmpty()) {
            return new SensitiveDataResult(0, Collections.emptyList());
        }

        String fullText = (subject != null ? subject + " " : "") + emailBody;
        String lowerText = fullText.toLowerCase();
        List<String> signals = new ArrayList<>();
        double score = 0;

        // Check regex patterns
        for (Map.Entry<String, Pattern> entry : PATTERNS.entrySet()) {
            Matcher matcher = entry.getValue().matcher(fullText);
            if (matcher.find()) {
                String signalName = entry.getKey() + "_exposed";
                signals.add(signalName);
                
                // Specific scores from request
                double signalScore = switch(entry.getKey()) {
                    case "ssn" -> 60;
                    case "credit_card" -> 70;
                    case "bank_account" -> 65;
                    default -> 25;
                };
                
                score += signalScore;
                log.debug("[SENSITIVE] Detected: {} score={}", signalName, signalScore);
            }
        }

        // Check financial keywords
        for (String keyword : FINANCIAL_KEYWORDS) {
            if (lowerText.contains(keyword)) {
                signals.add("financial_keyword_" + keyword.replace(" ", "_"));
                score += 10;
            }
        }

        // Cap at 100
        score = Math.min(score, 100);
        log.info("[SENSITIVE] Score={}, signals={}", score, signals.size());

        return new SensitiveDataResult(score, signals);
    }

    public record SensitiveDataResult(double score, List<String> signals) {}
}
